### PR TITLE
Qr Code

### PR DESCRIPTION
**Pull request by Princekumarofficial**
QR Code Scan Facility

### CODE REVIEW

File: api/views.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+from qrscan.models import MessCard, Meal
+from .permissions import IsStaff
+from .serializers import (
+    AllocationSerializer, PeriodSerializer, AllocationDetailSerializer, 
+    QRVerifySerializer, MealSerializer, UserSerializer, QRVerifyPostSerializer
+)
+from .utils.rebate_checker import is_student_on_rebate
+
+# Constants
+BREAKFAST_END_TIME = timezone.datetime.strptime('11:00:00', '%H:%M:%S').time()
```
Comment (mittal-ishaan) [inline api/views.py:19]: need to be admin panel editable.

File: messWebsite/settings.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = True
 
-ALLOWED_HOSTS = ["diningfee.iiti.ac.in", "127.0.0.1", "103.159.214.171"]
+ALLOWED_HOSTS = ["diningfee.iiti.ac.in", "127.0.0.1", "103.159.214.171", "*"]
```
Comment (mittal-ishaan) [inline messWebsite/settings.py:23]: this can be reverted

File: qrscan/admin.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    list_display = ['id', 'mess_card', 'date', 'breakfast', 'lunch', 'dinner']
+    list_display_links = ['id', 'mess_card']
+    search_fields = ('mess_card__allocation__student_id', 'mess_card__allocation__email__email')
+    list_filter = ('mess_card__allocation__caterer__name', 'date', 'breakfast', 'lunch', 'dinner')
+    raw_id_fields = ('mess_card',)
+    ordering = ['date', 'id']
+    list_per_page = 25
+
+# Registering models with the admin site
+admin.site.register(MessCard, MessCardAdmin)
```
Comment (mittal-ishaan) [inline qrscan/admin.py:25]: we can use decorators instead of these. Check already present admin.py for code consistency

File: qrscan/models.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+from home.models import Allocation, Student
+
+
+def generate_secret_key():
+    return secrets.token_hex(32)
+
+
+class MessCard(models.Model):
+    id = models.UUIDField(
```
Comment (mittal-ishaan) [inline qrscan/models.py:18]: check this https://github.com/pshrmn/notes/blob/master/django/models.md#primary-keys
maybe we might not need id

File: qrscan/models.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        qr.add_data(data)
+        qr.make(fit=True)
+
+        img = qr.make_image(fill="black", back_color="white").convert('RGB')
+
+        try:
+            logo = Image.open('./static/images/iiti_bw.png')
+            logo = logo.resize((150, 150))
+            pos = ((img.size[0] - logo.size[0]) // 2, (img.size[1] - logo.size[1]) // 2)
+            img.paste(logo, pos)
```
Comment (mittal-ishaan) [inline qrscan/models.py:67]: why are we doing this?

File: qrscan/models.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        if self.allocation and not self.student:
+            self.student = self.allocation.email
+        if not self.pk:
+            super().save(*args, **kwargs)
+        if not self.qr_code:
+            self.generate_qr_code()
+        super().save(*args, **kwargs)
+
+
+class Meal(models.Model):
```
Comment (mittal-ishaan) [inline qrscan/models.py:89]: also add an option for high tea. Just in case we may have it in future

File: qrscan/templates/mess_card.html [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      <span class="card-text">Hostel and Room: {{ student.hostel }} {{ student.room_no }}</span><br />
+      <div class="qr-code">
+        <img
+          alt="QR Code"
+          height="auro"
+          src="{{ mess_card.qr_code.url }}"
+          width="100%"
+        />
+      </div>
+      <div class="website">www.reallygreatsite.com</div>
```
Comment (mittal-ishaan) [inline qrscan/templates/mess_card.html:38]: maybe correction needed here

File: requirements.txt
Comment (mittal-ishaan) [inline requirements.txt:1]: use https://pypi.org/project/pipreqs/ to generate this file

**Review by mittal-ishaan — COMMENTED** [review]
few suggestions, but we can discuss on these after we go through the design once

File: qrscan/models.py (hunk 4 above)
Comment (Princekumarofficial) [reply qrscan/models.py:18]: It's important as anyone can generate an QR with a ID if its like Integer (1, 2, 3....)
Anyone can generate these numbers by guessing but by using this method nobody can do that

File: qrscan/models.py (hunk 5 above)
Comment (Princekumarofficial) [reply qrscan/models.py:67]: It adds iiti image to the center of the qr to look more professional

File: api/utils/rebate_checker.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def is_student_on_rebate(student: Student, allocation: Allocation):
+    """
+    This function checks if the student is on rebate or not
+    """
+    long_rebate = LongRebate.objects.filter(email=student)
+    today = timezone.localtime().date()
+
+    rebate_exists = TodayRebate.objects.filter(allocation_id=allocation.id).exists()
```
Comment (mittal-ishaan) [inline api/utils/rebate_checker.py:12]: TodayRebate has rebates that were applied on that day

File: api/utils/rebate_checker.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+from django.utils import timezone
+from home.models.students import Student, LongRebate, TodayRebate, Allocation
+
+
+def is_student_on_rebate(student: Student, allocation: Allocation):
+    """
+    This function checks if the student is on rebate or not
+    """
+    long_rebate = LongRebate.objects.filter(email=student)
```
Comment (mittal-ishaan) [inline api/utils/rebate_checker.py:9]: we can use less than and greater than checks while querying itself. this will reduce the in-memory items we will be having.
https://docs.djangoproject.com/en/5.1/ref/models/querysets/

File: api/views.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    View to update QR verification.
+    """
+    permission_classes = [IsStaff]
+
+    def _get_meal_type(self, time):
+        if time < BREAKFAST_END_TIME:
+            return 'breakfast'
+        elif time < LUNCH_END_TIME:
+            return 'lunch'
+        else:
```
Comment (mittal-ishaan) [inline api/views.py:113]: we might want to have strict lower and upper bound here for timings

File: api/views.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+                meal, _ = Meal.objects.get_or_create(mess_card=card, date=date)
+                is_rebate = is_student_on_rebate(card.student, card.allocation)
+
+                if card.allocation.period.end_date <= date:
+                    return Response({"success": False, "detail": "Not registered for current Period.", "mess_card": card_return_data}, status=status.HTTP_403_FORBIDDEN)
+
+                if card.allocation.caterer.name != request.user.username:
+                    return Response({"success": False, "detail": "Wrong caterer.", "mess_card": card_return_data}, status=status.HTTP_403_FORBIDDEN)
+
+                if is_rebate:
```
Comment (mittal-ishaan) [inline api/views.py:185]:
```suggestion
                if is_student_on_rebate(card.student, card.allocation):
```

File: api/views.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        today = timezone.localtime().date()
+        if meal_type == 'breakfast' and card.meal_set.filter(date=today, breakfast=True).exists():
+            return True
+        elif meal_type == 'lunch' and card.meal_set.filter(date=today, lunch=True).exists():
+            return True
+        elif meal_type == 'dinner' and card.meal_set.filter(date=today, dinner=True).exists():
+            return True
+        return False
+
+    def get(self, request):
```
Comment (mittal-ishaan) [inline api/views.py:138]: why is this required?

File: qrscan/models.py (hunk 4 above)
Comment (mittal-ishaan) [reply qrscan/models.py:18]: lets discuss on this.

File: qrscan/models.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+        buffer = BytesIO()
+        img.save(buffer, format='PNG')
+        file_name = f"qr_{self.student.email}_{self.allocation.student_id}.png"
+
+        self.qr_code.save(file_name, File(buffer), save=False)
+        buffer.close()
+
+    def save(self, *args, **kwargs):
+        if self.allocation and not self.student:
```
Comment (mittal-ishaan) [inline qrscan/models.py:80]: why?

**Comment by mittal-ishaan** [comment]
Right now: Caterer scans QR -> card validation done by portal -> card entry saved and student can eat
Suggested flow: Caterer scans QR -> card validation done by portal -> shows the result to caterer if successful then user details with student image if error then show error -> caterer goes through student picture and details and verifies the card -> card entry saved and student can eat.
thoughts? though this can be done in future PRs

File: api/views.py (hunk 12 above)
Comment (Princekumarofficial) [reply api/views.py:138]: not needed now, but can be used in future, like syncing data on caterer's device.

File: qrscan/models.py (hunk 13 above)
Comment (Princekumarofficial) [reply qrscan/models.py:80]: Error handling, in case self.student is not filled.

Comment (mittal-ishaan) [reply qrscan/models.py:80]: but we have on delete cascade right? will there be a case when `self.student` is `None`?

**Review by mittal-ishaan — APPROVED** [review]
good work!
